src/main/java/ch/njol/skript/bukkitutil/NamespacedUtils.java

```diff
@@ -0,0 +1,163 @@
+					// to represent the end of the hex sequence
+					if (i != (keyLength - 1) && LEGAL_NAMESPACE_CHARS.contains(key.charAt(i + 1)))
+						encodedKeyBuilder.append('.');
+				} else {
+					// we are dealing with a legal character, so we can just append it
+					encodedKeyBuilder.append(currentChar);
+				}
+			} else {
+				// add ".x(hex code)" to the encoded key
+				encodedKeyBuilder.append(".x");
+				encodedKeyBuilder.append(Integer.toHexString(currentChar));
+				// only add the trailing dot if the next character is a legal character
+				if (i != (keyLength - 1) && LEGAL_NAMESPACE_CHARS.contains(key.charAt(i + 1)))
+					encodedKeyBuilder.append('.');
+			}
+		}
+		if (skriptNamespace)
+			return NamespacedKey.fromString(encodedKeyBuilder.toString(), Skript.getInstance());
+		return NamespacedKey.fromString(encodedKeyBuilder.toString());
+	}
+
+	/**
+	 * Encodes a key to allow invalid character usage within the key.
+	 * If {@code key} contains a namespace identifier within, that will be the namespace.
+	 * Otherwise, it will create the key in Skript's namespace.
+	 *
+	 * @param key The key to use
+	 * @return The resulting {@link NamespacedKey} with an encoded key
+	 */
+	public static NamespacedKey createNamespacedKey(@NotNull String key) {
+		return createNamespacedKey(key, true);
+	}
+
+	/**
+	 * Decodes a NamespacedKey encoded by #getNamespacedKey
+	 *
+	 * @param namespacedKey the namespaced key to decode
+	 * @return a Pair with the first element as the namespace and the second as the decoded key
+	 */
+	public static NonNullPair<String, String> decodeNamespacedKey(NamespacedKey namespacedKey) {
+		String encodedKey = namespacedKey.getKey();
+		StringBuilder decodedKeyBuilder = new StringBuilder();
+		int encodedKeyLength = encodedKey.length();
+		int lastCharIndex = encodedKeyLength - 1;
+		for (int i = 0; i < encodedKeyLength; i++) {
+			char currentChar = encodedKey.charAt(i);
+			// if we are encountering a ".x" hex sequence
+			if (i != lastCharIndex && currentChar  == '.' && encodedKey.charAt(i + 1) == 'x') {
+				i += 2; // skip the ".x" so it isn't part of our hex string
+				StringBuilder hexString = new StringBuilder();
+				// The hex sequence continues until a . is encountered or we reach the end of the string
+				while (i <= lastCharIndex && encodedKey.charAt(i) != '.') {
+					hexString.append(encodedKey.charAt(i));
+					i++;
+				}
+				// if the . was the start of another ".x" hex sequence, back up by 1 character
+				if (i <= lastCharIndex && encodedKey.charAt(i + 1) == 'x')
+					i--;
+				// parse the hex sequence to a char
+				char decodedChar = (char) Long.parseLong(hexString.toString(), 16);
+				decodedKeyBuilder.append(decodedChar);
+			} else {
+				// this is just a normal character, not a hex sequence
+				// so we can just append it
+				decodedKeyBuilder.append(currentChar);
+			}
+		}
+		return new NonNullPair<>(namespacedKey.getNamespace(), decodedKeyBuilder.toString());
+	}
+
+}
```

Comment:
```suggestion
}

```

Reply:
Thank you kind sir, what would I do without you